### ISSUE TITLE
Feat: Ability to parse formulas for highlighting

### ISSUE THREAD

**Issue by MartinDawson** [handsontable/hyperformula#854]
So I'm doing formula syntax highlighting in my spreadsheet like this example:

![image](https://user-images.githubusercontent.com/15030491/139579297-4d8ad3d6-e826-4591-8ab9-b0ab94dc3755.png)

Currently hyperformula does formula parsing internally however there's no way to access it and the indexes in the string for it.

I tried to internally access lexer to do it but that doesn't work as it hasn't been parsed in the AST yet.

I also looked at `getCellPrecedents`, however this requires the cells to have been set already and in the dependency tree I think.

So something like:

@helper~
`parseFormulas(text: string)`: string[].

Would be great.

**Comment by MartinDawson** [comment]
If anyone knows of a work around to do this in the meantime that would be great as well.

**Comment by MartinDawson** [comment]
Actually. I have seen that `LexerConfig.js` holds the regexp patterns. So I can use these to find the cell references through text and then conver them to `SimpleCellAddresses` which I think I should be able to do with `simpleCellAddressFromString()`.

So nevermind on this feature.

Thanks!

**Comment by MartinDawson** [comment]
This is what I have got working for now:

```js
    const lexer = this.spreadsheet.hyperformula._parser.lexer;

    const { tokens } = lexer.tokenizeFormula(text);

    const cellReferenceParts: ICellReferencePart[] = [];
    const tokenParts = [];

    for (const [index, token] of tokens.entries()) {
      if (index === 0 && token.tokenType.name !== EqualsOp.name) {
        break;
      }

      if (token.tokenType.name === CellReference.name) {
        const color = getSyntaxColor();

        cellReferenceParts.push({
          startOffset: token.startOffset,
          endOffset: token.endOffset,
          referenceText: token.image,
          color,
        });
      }
    }
```

![image](https://user-images.githubusercontent.com/15030491/139952657-fdaca0dc-4958-4069-9f94-dc267a2d0f27.png)

You can see it incorrectly highlights `A2` in `2A2` even though it's invalid.

I'm thinking the best way to fix this is to iterate all characters before & after a cell reference up to the `Multiply/Add/Divide` etc Operations and see if each character is valid or not.

Then I will try put this in hyperformula as a `parser()` function as a PR..

**Comment by AMBudnik** [comment]
It looks great @MartinDawson !
do you use Handsontable as the table solution?

**Comment by MartinDawson** [comment]
@AMBudnik Thanks.

No we just use Hyperformula.js. We are creating our own which will match Google Sheets as close as possible because we needed a canvas based solution in the long term for our main application and not DOM.

It's still very early though with some bugs.

Here's an example demo: https://storybook.powersheet.io/?path=/story/spreadsheet--formulas

**Comment by AMBudnik** [comment]
Ah.. yes, understandable. 

I love this idea and I'm delighted to see such advanced use of HyperFormula.

**Comment by MartinDawson** [comment]
Yeah we use pretty much every feature of HF :). It would be impossible to do this without it (I tried & failed at creating my own parser before I found HF..) so it's really great.

**Comment by wojciechczerniak** [comment]
HF parser will fail with partially written formula:

![Screenshot 2021-11-03 at 14 25 10](https://user-images.githubusercontent.com/4718998/140068175-af4ed563-c19c-420d-87d1-b478fcc9842e.png)

You will not get any tokens for this scenario.

I was considering such a method when designing public API. But HF will not be very helpful here and I don't think this is the right way to go. HF will parse only full, valid formula string. Anything else is rejected. Those regexes are not ready for partially written formula, which is what we get during the editing process.

But you should have everything that is needed to implement such functionality.

---

Named expressions autocomplete can be achieved with [` listNamedExpressions`](https://handsontable.github.io/hyperformula/api/classes/hyperformula.html#listnamedexpressions) or [` getAllNamedExpressionsSerialized()`](https://handsontable.github.io/hyperformula/api/classes/hyperformula.html#getallnamedexpressionsserialized). So whenever you get anything that is not function, reference or string, you may filter the list to find best matches. Like GS does:

![Screenshot 2021-11-03 at 13 57 18](https://user-images.githubusercontent.com/4718998/140067563-9a54da4d-d6de-476e-89ac-727815ea7611.png)

---

For function list intelisense, like GS functionality:

![Screenshot 2021-11-03 at 13 57 27](https://user-images.githubusercontent.com/4718998/140065356-c3e17f6e-1fc8-42d2-b52b-f49fac809542.png)

You can use [` getRegisteredFunctionNames()`](https://handsontable.github.io/hyperformula/api/classes/hyperformula.html#getregisteredfunctionnames) method. Then filter the list

---

Function arguments intelisense may require some additional meta data storage #2. 

![Screenshot 2021-11-03 at 13 57 45](https://user-images.githubusercontent.com/4718998/140065182-ad312261-51b0-4baa-a3fa-351f8e11fdf1.png)

But those files should be stored next to the HF anyway and does not require any changes in HF. Of course, they are needed only if we need additional descriptions and guides (as well translations for them). Some basic argument intelisense should be possible with [` getAllFunctionPlugins()`](https://handsontable.github.io/hyperformula/api/classes/hyperformula.html#getallfunctionplugins) by reading meta from plugin static property `implementedFunctions`. Argument list and types are there for taking.

https://github.com/handsontable/hyperformula/blob/b989cfde24024309eae32a3edb069de84539ddeb/src/interpreter/plugin/SumifPlugin.ts#L63-L71

---

And if you want to get helpful tooltip with calculated value:

![Screenshot 2021-11-03 at 14 43 03](https://user-images.githubusercontent.com/4718998/140071155-de1d6568-d86f-472a-aa00-487afb1f5797.png)

use [`calculateFormula()`](https://handsontable.github.io/hyperformula/api/classes/hyperformula.html#calculateformula)

---

You probably already know, but for anyone else that's reading this "guide". When we have parsed range or cell reference with regex we can get cells coordinates to highglight with [`simpleCellAddressFromString()`](https://handsontable.github.io/hyperformula/api/classes/hyperformula.html#simplecelladdressfromstring)

![Screenshot 2021-11-03 at 14 45 47](https://user-images.githubusercontent.com/4718998/140071555-85454158-05d0-47aa-85ce-3262e2ba4d7e.png) or [`simpleCellRangeFromString()`](https://handsontable.github.io/hyperformula/api/classes/hyperformula.html#simplecellrangefromstring)

---

There are multiple partially written references `A:`, `Sheet1!`, `A1:B`, `NameExpr` when HF will fail. But formula editor experience should provide valid options to continue user input. Those need to be handled very carefully by custom regexes.

**Comment by MartinDawson** [comment]
@wojciechczerniak Got it,.

Yeah I was presuming that HF already had all this parsing internally and we could just re-use it or the regexes but I was wrong as it errors out if invalid strings are in the formula string (by design).

So custom regexes are the way to go here.

Thanks!

**Comment by MartinDawson** [comment]
Hmm the regex for all possible matches of cell references seems extremely complex as per this issue and very buggy: https://stackoverflow.com/questions/1897906/is-this-the-regex-for-matching-any-cell-reference-in-an-excel-formula

The fact that the regexes are split up and parsed individually through lexer makes it much easier...

I am thinking that using lexer/HF is still the way to go.

**Comment by wojciechczerniak** [comment]
I guess they are designed for different use case. I had a very custom solution in mind. 

Of course this can be done step by step. You can even use Chevrotain as HF does. This dependency would not add any weight to your app as it's already loaded by HyperFormula.

Syntax is well documented by [OpenDocument Standard, part 4: OpenFormula, in Section 5: Expression Syntax](https://docs.oasis-open.org/office/OpenDocument/v1.3/csprd02/part4-formula/OpenDocument-v1.3-csprd02-part4-formula.html#__RefHeading__1017930_715980110)

**Comment by crystalstone** [comment]
> 

hi，dear，I want to know whether your spreadsheet is open source and we can use

**Comment by AMBudnik** [comment]
Hi @crystalstone 

HyperFormula is dual-licensed. We have GPLv3 license for non-commercial use and for commercial use, it is under a Commercial License. More details https://handsontable.github.io/hyperformula/guide/licensing.html 

If you are not sure if you use case is commercial or not or have any other questions please feel free to contact our Sales Team at sales@handsontable.com

**Comment by BrianHung** [comment]
> You can even use Chevrotain as HF does.

I would use an incremental parser that's also error insensitive like [lezer](https://lezer.codemirror.net/). Syntax highlighting for parsing incomplete formulas is very similar to highlighting for incomplete code.

CodeMirror 5 previously had something for spreadsheets https://codemirror.net/mode/spreadsheet/ but that's outdated.

**Comment by sequba** [comment]
Formula syntax highlighting is a common use case for apps using HyperFormula. We'll analyze different solutions to make implementing this feature easier. As a result, we'll either:
- add API that can parse formulas, or
- create an official example that demonstrates the recommended approach to implementing custom formula parsing
